pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -0,0 +1,656 @@
+        if (StringUtils.isNotEmpty(brokerZnodePath)) {
+            try {
+                pulsar.getZkClient().delete(brokerZnodePath, -1);
+            } catch (Exception e) {
+                throw new PulsarServerException(e);
+            }
+        }
+    }
+
+    private String getNamespaceNameFromBundleName(String bundleName) {
+        // the bundle format is property/cluster/namespace/0x00000000_0xFFFFFFFF
+        int pos = bundleName.lastIndexOf("/");
+        checkArgument(pos != -1);
+        return bundleName.substring(0, pos);
+    }
+
+    private String getBundleRangeFromBundleName(String bundleName) {
+        // the bundle format is property/cluster/namespace/0x00000000_0xFFFFFFFF
+        int pos = bundleName.lastIndexOf("/");
+        checkArgument(pos != -1);
+        return bundleName.substring(pos + 1, bundleName.length());
+    }
+
+    /**
+     * As the leader broker, select bundles for the namespace service to unload so that they may be reassigned to new
+     * brokers.
+     */
+    @Override
+    public void doLoadShedding() {
+        for (LoadSheddingStrategy strategy : loadSheddingPipeline) {
+            final Map<String, String> bundlesToUnload = strategy.selectBundlesForUnloading(loadData, conf);
+            if (bundlesToUnload != null && !bundlesToUnload.isEmpty()) {
+                try {
+                    for (Map.Entry<String, String> entry : bundlesToUnload.entrySet()) {
+                        final String bundle = entry.getKey();
+                        final String broker = entry.getValue();
+                        adminCache.get(broker).namespaces().unloadNamespaceBundle(
+                                getNamespaceNameFromBundleName(bundle), getBundleRangeFromBundleName(bundle));
+                    }
+                } catch (Exception e) {
+                    log.warn("Error when trying to perform load shedding: {}", e);
+                }
+                return;
+            }
+        }
+    }
+
+    /**
+     * As the leader broker, attempt to automatically detect and split hot namespace bundles.
+     */
+    @Override
+    public void doNamespaceBundleSplit() {
+        // TODO?
+    }
+
+    @Override
+    public String getBrokerRoot() {
+        return LOADBALANCE_BROKERS_ROOT;
+    }
+
+    /**
+     * When the broker data ZooKeeper nodes are updated, update the broker data map.
+     */
+    @Override
+    public void onUpdate(final String path, final LocalBrokerData data, final Stat stat) {
+        scheduler.submit(this::updateAll);
+    }
+
+    /**
+     * As the leader broker, find a suitable broker for the assignment of the given bundle.
+     * 
+     * @param serviceUnit
+     *            ServiceUnitId for the bundle.
+     * @return The name of the selected broker, as it appears on ZooKeeper.
+     */
+    @Override
+    public synchronized String selectBrokerForAssignment(final ServiceUnitId serviceUnit) {
```

Comment:
Without the synchronized block, decisions will be made without the latest preallocated data. If many selections are processed within a short time frame, this could result in the least loaded broker receiving too many assignments.